tests for streaming encryption using arbitrary 10M byte strings

tomtitchener opened the pull request:
Tests for stream encrypt/decrypt of arbitrary 10M byte strings.

adinapoli commented:
Hey Tom, thanks for this PR!

I'm currently travelling and I will have a bit of a hectic weekend, so I'm
currently on mobile, so forgive the hasty tone!

This looks great, but there are a couple of things I would change before
merging.

In particular:

+ it feels weird to have this artificial "User10Mb" type, used only for
testing, in the library types. I suggest moving it inside the testing code,
so you don't need the bytestring-arbitrary dependency on the library side

+ better (aka seems a tad more idiomatic) to keep the resolver and
"install-ghc" entries at the top of the stack.yaml

Apart from.this I think the rest should be fine ;)

On 30 Sep 2016 10:12 p.m., "tomtitchener" <notifications@github.com> wrote:

> Tests for stream encrypt/decrypt of arbitrary 10M byte strings.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/RNCryptor/rncryptor-hs/pull/17
> Commit Summary
>
>    - tests for streaming encryption using arbitrary 10M byte strings
>
> File Changes
>
>    - *M* rncryptor.cabal
>    <https://github.com/RNCryptor/rncryptor-hs/pull/17/files#diff-0> (3)
>    - *M* src/Crypto/RNCryptor/Types.hs
>    <https://github.com/RNCryptor/rncryptor-hs/pull/17/files#diff-1> (29)
>    - *M* stack.yaml
>    <https://github.com/RNCryptor/rncryptor-hs/pull/17/files#diff-2> (9)
>    - *M* test/Main.hs
>    <https://github.com/RNCryptor/rncryptor-hs/pull/17/files#diff-3> (3)
>    - *M* test/Tests.hs
>    <https://github.com/RNCryptor/rncryptor-hs/pull/17/files#diff-4> (26)
>
> Patch Links:
>
>    - https://github.com/RNCryptor/rncryptor-hs/pull/17.patch
>    - https://github.com/RNCryptor/rncryptor-hs/pull/17.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RNCryptor/rncryptor-hs/pull/17>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAa-s9bGJel8DSavKtRxiOwDV0Os5k7Bks5qvW03gaJpZM4KLd_1>
> .
>

tomtitchener commented:
Thanks for the review, @adinapoli!  I moved the User10Mb type along with the other test specific types from Types.hs to Tests.hs which let me remove the bytestring-arbitrary from the library.  I updated stack.yaml to put the resolver and install-ghc lines at the top of the file.

When you get the time, please have a look at streamingRoundTrip in Tests.hs.  Connecting an output stream to an input stream through the buffer answered by listOutputStream feels awkward.  But I didn't find any library routine equivalent to System.IO.Streams.connect that plugged an output stream into an input stream.  I wonder if there's something fundamental here about streams that I don't understand.

adinapoli commented on stack.yaml:
```diff
@@ -1,6 +1,9 @@
+system-ghc: false
 extra-deps:
+- bytestring-arbitrary-0.1.0
+- text-1.2.2.1
```
Any reason why you need text listed as `extra-deps` here? When a library is already present in a Stackage snapshot (which is the case for `text`) you don't need to list it here explicitly. It will only make builds less stable, paradoxically, as it will impose yet another constraint to the cabal solver 😉

adinapoli commented:
@tomtitchener Re my answer above on your question, apparently GH does not allow me to edit a review text, which had a typo which makes the answer confusing, sorry. It really should be:

> I think this is by design (but I might be wrong, of course). I guess the idea is that an InputStream can ~~also~~ **only** produce values whereas an OutputStream can only consume them. I guess these would be equivalent to pipe's Producer and Consumer, which have very distinctive semantics which do not overlap. Looks like what you really wanted here is something like a Pipe, which is a Consumer and Producer rolled into one, but AFAIK io-stream does not expose such abstraction 😉

tomtitchener commented:
@adinapoli I removed text from extra deps and ran ran stack clean/stack build and stack clean/stack test successfully.

And re: not implementing pipe by design, wouldn't your argument apply equally well to the connect API?  Why have one but not the other, I wonder?  Usually there's a reason.  I just don't see it here.  Maybe something to do with buffering or laziness?  I wouldn't be surprised to learn my use of listOutputStream has some maximally-bad memory consumption behavior or something like that.

I did look into the pipes package but decided it wasn't worth the effort, given you can use existing stream APIs to achieve the same effect.  Unless I'm just clueless about streams and there's a really, really good reason for not using listOutputStream the way I do, which is what I worry about.

adinapoli commented:
> Why have one but not the other, I wonder?

Good question! I will have a think about that and I will get back to you when I'm less tired 😁 

> Unless I'm just clueless about streams and there's a really, really good reason for not using listOutputStream the way I do, which is what I worry about.

From a first look seems like `listOutputStream` provides you with that flush function as an "escape hatch" to go and collect the results. The library itself (if you look at the implementation) is very imperative and "side-effectful", but is also very fast and simple, which was my main rationale behind using it. So I don't think you should get any particularly bad memory profile/consumption, provided you use the "escape hatch" with care, as the usual caveats will apply: if you load a huge chunk of your stream into memory all at once, bad things could happen 😉
